vw/src/main/scala/com/microsoft/azure/synapse/ml/vw/VowpalWabbitCSETransformer.scala

```diff
@@ -0,0 +1,208 @@
+            .alias(Snips),
+          F.expr(s"ips($ProbabilityLoggedColName, ${rewardCol.col}, $ProbabilityPredictedColName, count)")
+            .alias(Ips),
+          F.expr(s"cressieRead($ProbabilityLoggedColName, ${rewardCol.col}, $ProbabilityPredictedColName, count, " +
+            s"$minImportanceWeight, $maxImportanceWeight)")
+            .alias(CressieRead),
+          F.expr(s"cressieReadInterval($ProbabilityLoggedColName, ${rewardCol.col}, " +
+            s"$ProbabilityPredictedColName, count, " +
+            s"$minImportanceWeight, $maxImportanceWeight, ${rewardCol.minRewardCol}, ${rewardCol.maxRewardCol})")
+            .alias(CressieReadInterval),
+          F.expr(s"cressieReadIntervalEmpirical($ProbabilityLoggedColName, ${rewardCol.col}, " +
+            s"$ProbabilityPredictedColName, count, " +
+            s"$minImportanceWeight, $maxImportanceWeight, ${rewardCol.minRewardCol}, ${rewardCol.maxRewardCol})")
+            .alias(CressieReadIntervalEmp)
+        ).alias(rewardCol.name)
+      }
+      })
+
+  override def transform(dataset: Dataset[_]): DataFrame = {
+    logTransform[DataFrame]({
+      val df = dataset.toDF
+
+      // fetch reward columns out of nested structured
+      val rewardsCol = rewardColumns(df.schema)
+
+      // calculate min/max for each reward
+      val minMaxes = rewardsCol
+        .flatMap({ rewardCol =>
+          Seq(F.min(rewardCol.col).alias(rewardCol.minRewardCol),
+            F.max(rewardCol.col).alias(rewardCol.maxRewardCol))
+        })
+
+      val minMaxRewards = df.agg(minMaxes.head, minMaxes.drop(1): _*)
+
+      // register contextual bandit policy evaluation aggregates
+      PolicyEvalUDAFUtil.registerUdafs()
+
+      val metrics = globalMetrics ++ perRewardMetrics(rewardsCol, getMinImportanceWeight, getMaxImportanceWeight)
+
+      df.crossJoin(minMaxRewards.hint("broadcast"))
+        // probability predicted = 1 if top action matches observed action, otherwise 0
+        // Note: using the probability distribution produces by cb_adf_explore only increases variance
+        // and doesn't help model selection. also the offline policy can't influence what data is getting
+        // collected offline
+        .withColumn(ProbabilityPredictedColName,
+          F.when(F.expr(s"element_at(predictions, 1).action == $ChosenActionIndexColName"), 1f)
+            .otherwise(0f))
+        // example weight, defaults to 1
+        .withColumn("count", F.lit(1))
+        .withColumn("w", F.col(ProbabilityPredictedColName) / F.col(ProbabilityLoggedColName))
+        // optional stratification
+        .groupBy(getMetricsStratificationCols.map(F.col): _*)
+        .agg(metrics.head, metrics.drop(1): _*)
+    })
+  }
+
+  override def transformSchema(schema: StructType): StructType =
+    T.StructType(
+      // groupBy
+      getMetricsStratificationCols.map(T.StructField(_, T.StringType, true)) ++
+      Seq(
```

Comment:
maybe use case classes and the SparkBindings API to remove the need for manual schemas

Reply:
as discussed, due to dynamics cols not possible.